// @directus/shared/dist/cjs/utils/node/generate-extensions-entry.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const generate_extensions_entry_1 = require("./generate-extensions-entry");
(0, vitest_1.describe)('generateExtensionsEntry', () => {
    const type = 'panel';
    (0, vitest_1.it)('returns an extension entrypoint exporting all extensions with a type that matches the provided type', () => {
        const mockExtension = [
            { path: './extensions', name: 'mockExtension', type: 'panel', entrypoint: 'index.js', local: true },
        ];
        (0, vitest_1.expect)((0, generate_extensions_entry_1.generateExtensionsEntry)(type, mockExtension)).toBe(`import e0 from './extensions/index.js';
export default [e0];`);
    });
    (0, vitest_1.it)('returns an empty extension entrypoint if there is no extension with the provided type', () => {
        const mockExtension = [
            {
                path: './extensions',
                name: 'mockExtension',
                type: 'pack',
                version: '1.0.0',
                host: '^9.0.0',
                children: [],
                local: false,
            },
        ];
        (0, vitest_1.expect)((0, generate_extensions_entry_1.generateExtensionsEntry)(type, mockExtension)).toBe(`export default [];`);
    });
});
