// @directus/shared/dist/cjs/utils/define-extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineOperationApi = exports.defineOperationApp = exports.defineEndpoint = exports.defineHook = exports.definePanel = exports.defineModule = exports.defineLayout = exports.defineDisplay = exports.defineInterface = void 0;
function defineInterface(config) {
    return config;
}
exports.defineInterface = defineInterface;
function defineDisplay(config) {
    return config;
}
exports.defineDisplay = defineDisplay;
function defineLayout(config) {
    return config;
}
exports.defineLayout = defineLayout;
function defineModule(config) {
    return config;
}
exports.defineModule = defineModule;
function definePanel(config) {
    return config;
}
exports.definePanel = definePanel;
function defineHook(config) {
    return config;
}
exports.defineHook = defineHook;
function defineEndpoint(config) {
    return config;
}
exports.defineEndpoint = defineEndpoint;
function defineOperationApp(config) {
    return config;
}
exports.defineOperationApp = defineOperationApp;
function defineOperationApi(config) {
    return config;
}
exports.defineOperationApi = defineOperationApi;
