// @directus/shared/dist/cjs/utils/get-collection-type.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const get_collection_type_1 = require("./get-collection-type");
const TableCollection = {
    collection: 'table',
    schema: {
        name: 'table',
    },
    meta: null,
};
const AliasCollection = {
    collection: 'table',
    schema: null,
    meta: {
        collection: 'table',
        note: '',
        hidden: true,
        singleton: true,
        icon: 'box',
        color: '#abcabc',
        translations: null,
        display_template: null,
        sort_field: null,
        archive_field: null,
        archive_value: null,
        unarchive_value: null,
        archive_app_filter: true,
        item_duplication_fields: null,
        accountability: null,
        sort: null,
        group: null,
        collapse: 'open',
    },
};
const UnknownCollection = {
    collection: 'unknown',
    schema: null,
    meta: null,
};
(0, vitest_1.describe)('getCollectionType', () => {
    (0, vitest_1.it)('returns "table" when collection has schema information', () => {
        (0, vitest_1.expect)((0, get_collection_type_1.getCollectionType)(TableCollection)).toStrictEqual('table');
    });
    (0, vitest_1.it)('returns "alias" when collection has schema information', () => {
        (0, vitest_1.expect)((0, get_collection_type_1.getCollectionType)(AliasCollection)).toStrictEqual('alias');
    });
    (0, vitest_1.it)('returns "unknown" when collection has schema information', () => {
        (0, vitest_1.expect)((0, get_collection_type_1.getCollectionType)(UnknownCollection)).toStrictEqual('unknown');
    });
});
