// directus/dist/database/helpers/fn/dialects/mssql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FnHelperMSSQL = void 0;
const types_1 = require("../types");
class FnHelperMSSQL extends types_1.FnHelper {
    year(table, column) {
        return this.knex.raw('DATEPART(year, ??.??)', [table, column]);
    }
    month(table, column) {
        return this.knex.raw('DATEPART(month, ??.??)', [table, column]);
    }
    week(table, column) {
        return this.knex.raw('DATEPART(week, ??.??)', [table, column]);
    }
    day(table, column) {
        return this.knex.raw('DATEPART(day, ??.??)', [table, column]);
    }
    weekday(table, column) {
        return this.knex.raw('DATEPART(weekday, ??.??)', [table, column]);
    }
    hour(table, column) {
        return this.knex.raw('DATEPART(hour, ??.??)', [table, column]);
    }
    minute(table, column) {
        return this.knex.raw('DATEPART(minute, ??.??)', [table, column]);
    }
    second(table, column) {
        return this.knex.raw('DATEPART(second, ??.??)', [table, column]);
    }
    count(table, column, options) {
        var _a, _b, _c, _d, _e;
        const type = (_e = (_d = (_c = (_b = (_a = this.schema.collections) === null || _a === void 0 ? void 0 : _a[table]) === null || _b === void 0 ? void 0 : _b.fields) === null || _c === void 0 ? void 0 : _c[column]) === null || _d === void 0 ? void 0 : _d.type) !== null && _e !== void 0 ? _e : 'unknown';
        if (type === 'json') {
            return this.knex.raw(`(SELECT COUNT(*) FROM OPENJSON(??.??, '$'))`, [table, column]);
        }
        if (type === 'alias') {
            return this._relationalCount(table, column, options);
        }
        throw new Error(`Couldn't extract type from ${table}.${column}`);
    }
}
exports.FnHelperMSSQL = FnHelperMSSQL;
