// @aws-sdk/client-ses/dist-es/commands/CreateCustomVerificationEmailTemplateCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateCustomVerificationEmailTemplateRequestFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryCreateCustomVerificationEmailTemplateCommand, serializeAws_queryCreateCustomVerificationEmailTemplateCommand, } from "../protocols/Aws_query";
export class CreateCustomVerificationEmailTemplateCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "CreateCustomVerificationEmailTemplateCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: CreateCustomVerificationEmailTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: (output) => output,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryCreateCustomVerificationEmailTemplateCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryCreateCustomVerificationEmailTemplateCommand(output, context);
    }
}
