// @aws-sdk/client-ses/dist-es/commands/CreateConfigurationSetTrackingOptionsCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateConfigurationSetTrackingOptionsRequestFilterSensitiveLog, CreateConfigurationSetTrackingOptionsResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryCreateConfigurationSetTrackingOptionsCommand, serializeAws_queryCreateConfigurationSetTrackingOptionsCommand, } from "../protocols/Aws_query";
export class CreateConfigurationSetTrackingOptionsCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "CreateConfigurationSetTrackingOptionsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: CreateConfigurationSetTrackingOptionsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: CreateConfigurationSetTrackingOptionsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryCreateConfigurationSetTrackingOptionsCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryCreateConfigurationSetTrackingOptionsCommand(output, context);
    }
}
