// @aws-sdk/client-ses/dist-es/commands/DeleteTemplateCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteTemplateRequestFilterSensitiveLog, DeleteTemplateResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryDeleteTemplateCommand, serializeAws_queryDeleteTemplateCommand, } from "../protocols/Aws_query";
export class DeleteTemplateCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "DeleteTemplateCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: DeleteTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DeleteTemplateResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryDeleteTemplateCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryDeleteTemplateCommand(output, context);
    }
}
