// @aws-sdk/client-ses/dist-es/commands/DeleteConfigurationSetCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteConfigurationSetRequestFilterSensitiveLog, DeleteConfigurationSetResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryDeleteConfigurationSetCommand, serializeAws_queryDeleteConfigurationSetCommand, } from "../protocols/Aws_query";
export class DeleteConfigurationSetCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "DeleteConfigurationSetCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: DeleteConfigurationSetRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DeleteConfigurationSetResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryDeleteConfigurationSetCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryDeleteConfigurationSetCommand(output, context);
    }
}
