// @aws-sdk/client-ses/dist-es/commands/DeleteReceiptRuleSetCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteReceiptRuleSetRequestFilterSensitiveLog, DeleteReceiptRuleSetResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryDeleteReceiptRuleSetCommand, serializeAws_queryDeleteReceiptRuleSetCommand, } from "../protocols/Aws_query";
export class DeleteReceiptRuleSetCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "DeleteReceiptRuleSetCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: DeleteReceiptRuleSetRequestFilterSensitiveLog,
            outputFilterSensitiveLog: DeleteReceiptRuleSetResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryDeleteReceiptRuleSetCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryDeleteReceiptRuleSetCommand(output, context);
    }
}
