// @aws-sdk/client-ses/dist-es/commands/GetAccountSendingEnabledCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetAccountSendingEnabledResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryGetAccountSendingEnabledCommand, serializeAws_queryGetAccountSendingEnabledCommand, } from "../protocols/Aws_query";
export class GetAccountSendingEnabledCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "GetAccountSendingEnabledCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: (input) => input,
            outputFilterSensitiveLog: GetAccountSendingEnabledResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryGetAccountSendingEnabledCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryGetAccountSendingEnabledCommand(output, context);
    }
}
