// @aws-sdk/client-ses/dist-es/commands/GetIdentityNotificationAttributesCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetIdentityNotificationAttributesRequestFilterSensitiveLog, GetIdentityNotificationAttributesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryGetIdentityNotificationAttributesCommand, serializeAws_queryGetIdentityNotificationAttributesCommand, } from "../protocols/Aws_query";
export class GetIdentityNotificationAttributesCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "GetIdentityNotificationAttributesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: GetIdentityNotificationAttributesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetIdentityNotificationAttributesResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryGetIdentityNotificationAttributesCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryGetIdentityNotificationAttributesCommand(output, context);
    }
}
