// @aws-sdk/client-ses/dist-es/commands/GetIdentityVerificationAttributesCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetIdentityVerificationAttributesRequestFilterSensitiveLog, GetIdentityVerificationAttributesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryGetIdentityVerificationAttributesCommand, serializeAws_queryGetIdentityVerificationAttributesCommand, } from "../protocols/Aws_query";
export class GetIdentityVerificationAttributesCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "GetIdentityVerificationAttributesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: GetIdentityVerificationAttributesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetIdentityVerificationAttributesResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryGetIdentityVerificationAttributesCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryGetIdentityVerificationAttributesCommand(output, context);
    }
}
