// @aws-sdk/client-ses/dist-es/commands/ListTemplatesCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListTemplatesRequestFilterSensitiveLog, ListTemplatesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryListTemplatesCommand, serializeAws_queryListTemplatesCommand, } from "../protocols/Aws_query";
export class ListTemplatesCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "ListTemplatesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: ListTemplatesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: ListTemplatesResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryListTemplatesCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryListTemplatesCommand(output, context);
    }
}
