// @aws-sdk/client-ses/dist-es/commands/PutConfigurationSetDeliveryOptionsCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutConfigurationSetDeliveryOptionsRequestFilterSensitiveLog, PutConfigurationSetDeliveryOptionsResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryPutConfigurationSetDeliveryOptionsCommand, serializeAws_queryPutConfigurationSetDeliveryOptionsCommand, } from "../protocols/Aws_query";
export class PutConfigurationSetDeliveryOptionsCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "PutConfigurationSetDeliveryOptionsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: PutConfigurationSetDeliveryOptionsRequestFilterSensitiveLog,
            outputFilterSensitiveLog: PutConfigurationSetDeliveryOptionsResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryPutConfigurationSetDeliveryOptionsCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryPutConfigurationSetDeliveryOptionsCommand(output, context);
    }
}
