// @aws-sdk/client-ses/dist-es/commands/SendRawEmailCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendRawEmailRequestFilterSensitiveLog, SendRawEmailResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_querySendRawEmailCommand, serializeAws_querySendRawEmailCommand } from "../protocols/Aws_query";
export class SendRawEmailCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "SendRawEmailCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: SendRawEmailRequestFilterSensitiveLog,
            outputFilterSensitiveLog: SendRawEmailResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_querySendRawEmailCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_querySendRawEmailCommand(output, context);
    }
}
