// @aws-sdk/client-ses/dist-es/commands/SetIdentityDkimEnabledCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SetIdentityDkimEnabledRequestFilterSensitiveLog, SetIdentityDkimEnabledResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_querySetIdentityDkimEnabledCommand, serializeAws_querySetIdentityDkimEnabledCommand, } from "../protocols/Aws_query";
export class SetIdentityDkimEnabledCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "SetIdentityDkimEnabledCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: SetIdentityDkimEnabledRequestFilterSensitiveLog,
            outputFilterSensitiveLog: SetIdentityDkimEnabledResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_querySetIdentityDkimEnabledCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_querySetIdentityDkimEnabledCommand(output, context);
    }
}
