// @aws-sdk/client-ses/dist-es/commands/UpdateTemplateCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateTemplateRequestFilterSensitiveLog, UpdateTemplateResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_queryUpdateTemplateCommand, serializeAws_queryUpdateTemplateCommand, } from "../protocols/Aws_query";
export class UpdateTemplateCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "UpdateTemplateCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: UpdateTemplateRequestFilterSensitiveLog,
            outputFilterSensitiveLog: UpdateTemplateResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryUpdateTemplateCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryUpdateTemplateCommand(output, context);
    }
}
