// @aws-sdk/client-ses/dist-es/commands/VerifyEmailAddressCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { VerifyEmailAddressRequestFilterSensitiveLog } from "../models/models_0";
import { deserializeAws_queryVerifyEmailAddressCommand, serializeAws_queryVerifyEmailAddressCommand, } from "../protocols/Aws_query";
export class VerifyEmailAddressCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SESClient";
        const commandName = "VerifyEmailAddressCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: VerifyEmailAddressRequestFilterSensitiveLog,
            outputFilterSensitiveLog: (output) => output,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_queryVerifyEmailAddressCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_queryVerifyEmailAddressCommand(output, context);
    }
}
