// @directus/schema/dist/dialects/sqlite.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var sqlite_1 = __importStar(require("knex-schema-inspector/dist/dialects/sqlite"));
var extract_max_length_1 = __importDefault(require("knex-schema-inspector/dist/utils/extract-max-length"));
var extract_type_1 = __importDefault(require("knex-schema-inspector/dist/utils/extract-type"));
var SQLite = /** @class */ (function (_super) {
    __extends(SQLite, _super);
    function SQLite() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SQLite.prototype.overview = function () {
        return __awaiter(this, void 0, void 0, function () {
            var tablesWithAutoIncrementPrimaryKeys, tables, overview, _i, tables_1, table, columns, primaryKeys, _a, columns_1, column;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.knex.select('name').from('sqlite_master').whereRaw("sql LIKE \"%AUTOINCREMENT%\"")];
                    case 1:
                        tablesWithAutoIncrementPrimaryKeys = (_b.sent()).map(function (_a) {
                            var name = _a.name;
                            return name;
                        });
                        return [4 /*yield*/, this.tables()];
                    case 2:
                        tables = _b.sent();
                        overview = {};
                        _i = 0, tables_1 = tables;
                        _b.label = 3;
                    case 3:
                        if (!(_i < tables_1.length)) return [3 /*break*/, 6];
                        table = tables_1[_i];
                        return [4 /*yield*/, this.knex.raw("PRAGMA table_xinfo(??)", table)];
                    case 4:
                        columns = _b.sent();
                        if (table in overview === false) {
                            primaryKeys = columns.filter(function (column) { return column.pk !== 0; });
                            overview[table] = {
                                primary: primaryKeys.length !== 1 ? undefined : primaryKeys[0].name,
                                columns: {},
                            };
                        }
                        for (_a = 0, columns_1 = columns; _a < columns_1.length; _a++) {
                            column = columns_1[_a];
                            overview[table].columns[column.name] = {
                                table_name: table,
                                column_name: column.name,
                                default_value: column.pk === 1 && tablesWithAutoIncrementPrimaryKeys.includes(table)
                                    ? 'AUTO_INCREMENT'
                                    : (0, sqlite_1.parseDefaultValue)(column.dflt_value),
                                is_nullable: column.notnull == 0,
                                is_generated: column.hidden !== 0,
                                data_type: (0, extract_type_1.default)(column.type),
                                max_length: (0, extract_max_length_1.default)(column.type),
                                numeric_precision: null,
                                numeric_scale: null,
                            };
                        }
                        _b.label = 5;
                    case 5:
                        _i++;
                        return [3 /*break*/, 3];
                    case 6: return [2 /*return*/, overview];
                }
            });
        });
    };
    return SQLite;
}(sqlite_1.default));
exports.default = SQLite;
