// directus/dist/operations/exec/index.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vm2_1 = require("vm2");
const index_1 = __importDefault(require("./index"));
test('Rejects when modules are used without modules being allowed', async () => {
    const testCode = `
		const test = require('test');
	`;
    await expect(index_1.default.handler({ code: testCode }, {
        data: {},
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).rejects.toEqual(new vm2_1.VMError("Cannot find module 'test'"));
});
test('Rejects when code contains syntax errors', async () => {
    const testCode = `
		~~
	`;
    await expect(index_1.default.handler({ code: testCode }, {
        data: {},
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).rejects.toEqual(new Error("Couldn't compile code: Unexpected end of input"));
});
test('Rejects when returned function does something illegal', async () => {
    const testCode = `
		module.exports = function() {
			return a + b;
		};
	`;
    await expect(index_1.default.handler({ code: testCode }, {
        data: {},
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).rejects.toEqual(new ReferenceError('a is not defined'));
});
test("Rejects when code doesn't return valid function", async () => {
    const testCode = `
		module.exports = false;
	`;
    await expect(index_1.default.handler({ code: testCode }, {
        data: {},
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).rejects.toEqual(new TypeError('fn is not a function'));
});
test('Rejects returned function throws errors', async () => {
    const testCode = `
		module.exports = function () {
			throw new Error('test');
		};
	`;
    await expect(index_1.default.handler({ code: testCode }, {
        data: {},
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).rejects.toEqual(new Error('test'));
});
test('Executes function when valid', () => {
    const testCode = `
		module.exports = function (data) {
			return { result: data.input + ' test' };
		};
	`;
    expect(index_1.default.handler({ code: testCode }, {
        data: {
            input: 'start',
        },
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: '',
        },
    })).resolves.toEqual({ result: 'start test' });
});
test('Allows modules that are whitelisted', () => {
    const testCode = `
		const bytes = require('bytes');

		module.exports = function (data) {
			return { result: bytes(1000) };
		};
	`;
    expect(index_1.default.handler({ code: testCode }, {
        env: {
            FLOWS_EXEC_ALLOWED_MODULES: 'bytes',
        },
    })).resolves.toEqual({ result: '1000B' });
});
