// @directus/shared/dist/esm/constants/extensions.js
export const APP_SHARED_DEPS = ['@directus/extensions-sdk', 'vue', 'vue-router', 'vue-i18n', 'pinia'];
export const API_SHARED_DEPS = ['directus'];
export const APP_EXTENSION_TYPES = ['interface', 'display', 'layout', 'module', 'panel'];
export const API_EXTENSION_TYPES = ['hook', 'endpoint'];
export const HYBRID_EXTENSION_TYPES = ['operation'];
export const EXTENSION_TYPES = [...APP_EXTENSION_TYPES, ...API_EXTENSION_TYPES, ...HYBRID_EXTENSION_TYPES];
export const PACKAGE_EXTENSION_TYPES = ['pack'];
export const EXTENSION_PACKAGE_TYPES = [...EXTENSION_TYPES, ...PACKAGE_EXTENSION_TYPES];
export const APP_OR_HYBRID_EXTENSION_TYPES = [...APP_EXTENSION_TYPES, ...HYBRID_EXTENSION_TYPES];
export const API_OR_HYBRID_EXTENSION_TYPES = [...API_EXTENSION_TYPES, ...HYBRID_EXTENSION_TYPES];
export const APP_OR_HYBRID_EXTENSION_PACKAGE_TYPES = [
    ...APP_OR_HYBRID_EXTENSION_TYPES,
    ...PACKAGE_EXTENSION_TYPES,
];
export const API_OR_HYBRID_EXTENSION_PACKAGE_TYPES = [
    ...API_OR_HYBRID_EXTENSION_TYPES,
    ...PACKAGE_EXTENSION_TYPES,
];
export const EXTENSION_LANGUAGES = ['javascript', 'typescript'];
export const EXTENSION_NAME_REGEX = /^(?:(?:@[^/]+\/)?directus-extension-|@directus\/extension-).+$/;
export const EXTENSION_PKG_KEY = 'directus:extension';
